analyzer/protocol/openvpn/main.zeek

```diff
@@ -0,0 +1,90 @@
+module OpenVPN;
+
+export {
+	## Set to true to disable the analyzer after the handshake is detected.
+	## This helps reduce processing if you will not look at all of the OpenVPN
+	## traffic.
+	option disable_analyzer_after_detection = T;
+
+	type ControlMsg: record {
+		## Opcode
+		opcode					:	count;
+		## Key ID
+		key_id					:	count;
+		## Session ID
+		session_id	    		: 	string &optional;
+		## Packet id ack array
+		packet_id_ack_array		:	vector of count &optional;
+		## Remote session ID
+		remote_session_id		:	string &optional;
+		## Packet ID
+		packet_id				:	count &optional;
+		## The amount of data
+		data_len				:	count;
+	};
+
+	type AckMsg: record {
+		## Opcode
+		opcode					:	count;
+		## Key ID
+		key_id					:	count;
+		## Session ID
+		session_id	    		: 	string &optional;
+		## Packet id ack array
+		packet_id_ack_array		:	vector of count &optional;
+		## Remote session ID
+		remote_session_id		:	string &optional;
+	};
+
+	type DataMsg: record {
+		## Opcode
+		opcode					:	count;
+		## Key ID
+		key_id					:	count;
+		## The amount of data
+		data_len				:	count;
+		## Peer ID
+		peer_id		    		: 	string &optional;
+	};
+
+	global OpenVPN::control_message: event(c: connection, is_orig: bool, msg: OpenVPN::ControlMsg);
+	global OpenVPN::data_message: event(c: connection, is_orig: bool, msg: OpenVPN::DataMsg);
+	global OpenVPN::ack_message: event(c: connection, is_orig: bool, msg: OpenVPN::AckMsg);
+
+	## The record type which contains OpenVPN info.
+	type Info: record {
+		## The analyzer ID used for the analyzer instance attached
+		## to each connection.  It is not used for logging since it's a
+		## meaningless arbitrary number.
+		analyzer_id:      count            &optional;
+	};
+
+}
+
+redef record connection += {
+	openvpn: Info &optional;
+};
+
+function set_session(c: connection)
+	{
+	if ( ! c?$openvpn )
+		c$openvpn = [];
+	}
+
+event protocol_confirmation(c: connection, atype: Analyzer::Tag, aid: count) &priority=5
+	{
+	set_session(c);
+	if ( atype == Analyzer::ANALYZER_SPICY_OPENVPN || atype == Analyzer::ANALYZER_SPICY_OPENVPNHMAC || atype == Analyzer::ANALYZER_SPICY_OPENVPNTCP || atype == Analyzer::ANALYZER_SPICY_OPENVPNTCPHMAC )
+		{
+		c$openvpn$analyzer_id = aid;
+		}
+	}
+
+event OpenVPN::data_message(c: connection, is_orig: bool, msg: OpenVPN::DataMsg)
+	{
+	if (disable_analyzer_after_detection == T && c?$openvpn && c$openvpn?$analyzer_id)
```

Comment:
Assuming we don't log anything else - besides the fact that the connection is OpenVPN - is there a reason to not just do the disable_analyzer() bit in the protocol_confirmation event?

That way you don't even need the new Info record.

Reply:
Yes, so researchers can leave everything enabled to grab things like sequence of lengths, then turn off the analyzer when they are done.  It's a specific piece of logic a researcher asked me for.

Reply:
Ok, that makes sense. 

Though - it honestly sounds like then they should just do the tracking of the analyzer id in their scripts. This way you don't add a global datastructure that is not needed by most people. :)

Reply:
That's probably true.  The original query was basically "can you make it behave in the same way ssl behaves" wrt to this functionality so I did.